File: frontend/mobile/lib/services/recommendation_service.dart
import 'package:dio/dio.dart';

import '../core/api_client.dart';
import '../models/api_error.dart';
import '../models/city.dart';
import '../models/responses/product_response.dart';
import '../models/responses/recommendation_response.dart';
import 'api_exception.dart';

/// Recommendation surfaces — mirrors `RecommendationService` in `frontend/web`. Same
/// `_isSuccess`/`_handleError` shape as [CatalogService], which is the convention every service in
/// this folder follows.
class RecommendationService {
  bool _isSuccess(int? code) => code != null && code >= 200 && code < 300;

  Never _handleError(Response response) {
    ApiError apiError;
    try {
      final body = response.data;
      apiError = body is Map<String, dynamic> ? ApiError.fromJson(body) : ApiError(message: body?.toString());
    } catch (_) {
      apiError = ApiError(message: response.data?.toString());
    }
    throw ApiException(statusCode: response.statusCode ?? 0, apiError: apiError);
  }

  /// GET /api/recommendations/me — requires a session. Never comes back empty for lack of
  /// history: the backend falls through to popularity and says so in `source`, which is what the
  /// row title is derived from.
  Future<RecommendationResponse> getForMe({int take = 8}) async {
    final response = await apiClient.get('Recommendations/me', queryParameters: {'take': take});
    if (!_isSuccess(response.statusCode)) _handleError(response);
    return RecommendationResponse.fromJson(response.data as Map<String, dynamic>);
  }

  /// GET /api/recommendations/similar/{productId} — public, needs no user history.
  Future<List<ProductResponse>> getSimilar(String productId, {int take = 6}) async {
    final response = await apiClient.get(
      'Recommendations/similar/$productId',
      queryParameters: {'take': take},
    );
    if (!_isSuccess(response.statusCode)) _handleError(response);
    return (response.data as List<dynamic>)
        .map((item) => ProductResponse.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  /// GET /api/recommendations/popular — public.
  Future<List<ProductResponse>> getPopular({City? city, int take = 8}) async {
    final response = await apiClient.get('Recommendations/popular', queryParameters: {
      'take': take,
      if (city != null) 'city': cityToJson(city),
    });
    if (!_isSuccess(response.statusCode)) _handleError(response);
    return (response.data as List<dynamic>)
        .map((item) => ProductResponse.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  /// POST /api/recommendations/views — fire-and-forget. Swallows every failure on purpose:
  /// recording a view is a background nicety, and nothing about someone opening a product should
  /// break because the recommender is unreachable.
  Future<void> trackView(String productId) async {
    try {
      await apiClient.post('Recommendations/views', data: {'productId': productId});
    } catch (_) {
      // Intentionally ignored — see doc comment.
    }
  }
}

File: frontend/web/src/app/core/services/recommendation.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable, map } from 'rxjs';

import { environment } from '../../../environments/environment';
import { City, Product, normalizeProduct } from '../models/catalog.models';
import {
  RecommendationResult,
  toRecommendationSource,
} from '../models/recommendation.models';

@Injectable({ providedIn: 'root' })
export class RecommendationService {
  private readonly http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiBaseUrl}/recommendations`;

  /**
   * GET /api/recommendations/me — signed-in visitors only. Never returns an empty list for lack
   * of history: the API falls back to popularity, and reports which strategy it used in `source`.
   */
  getForMe(take = 8): Observable<RecommendationResult> {
    return this.http
      .get<RecommendationResult>(`${this.baseUrl}/me`, { params: { take: String(take) } })
      .pipe(
        map((result) => ({
          items: result.items.map(normalizeProduct),
          source: toRecommendationSource(result.source),
        })),
      );
  }

  /** GET /api/recommendations/similar/{productId} — public, needs no user history. */
  getSimilar(productId: string, take = 6): Observable<Product[]> {
    return this.http
      .get<Product[]>(`${this.baseUrl}/similar/${productId}`, { params: { take: String(take) } })
      .pipe(map((products) => products.map(normalizeProduct)));
  }

  /** GET /api/recommendations/popular — public. What anonymous visitors see in place of a
   * personalized row. */
  getPopular(city?: City | null, take = 8): Observable<Product[]> {
    const params: Record<string, string> = { take: String(take) };
    if (city) params['city'] = city;

    return this.http
      .get<Product[]>(`${this.baseUrl}/popular`, { params })
      .pipe(map((products) => products.map(normalizeProduct)));
  }

  /**
   * POST /api/recommendations/views — fire-and-forget from the caller's point of view. Callers
   * must still subscribe (an unsubscribed HttpClient observable never fires) and must swallow the
   * error: failing to record a view is not something to interrupt someone's browsing over.
   */
  trackView(productId: string): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/views`, { productId });
  }
}
